designs/0028-parallel_map.md

```diff
@@ -0,0 +1,129 @@
+- Feature Name: parallel_map
+- Start Date: 2020/09/29
+- RFC PR: (leave this empty)
+- Stan Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+The proposed ```parallel_map``` functionality provides a means of within-chain parallelisation that does not require the sum of iterations be returned (i.e., ```reduce_sum```) or any complex argument-packing (i.e., ```map_rect```).
+
+# Motivation
+[motivation]: #motivation
+
+While the ```reduce_sum``` framework has been very popular for its ease of use, it's limited to situations where the goal is to return the sum of independent computations. Many users have models with large functions/computations that return vectors or matrices in full, rather than their sum. For these cases, users seeking parallelisation are limited to the ```map_rect``` functionality, which has a non-zero barrier to entry due to the need for argument packing and unpacking. The ```parallel_map``` framework aims to provide a means of parallelising these operations in a manner which is easy for users to understand and implement.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+## Target Loop
+As an example, let's say the user wanted to calculate the log of the beta function from two vectors of size 10000:
+```
+for (int i = 0; i < 10000; ++i) {
+  result[i] = lgamma(a[i]) + lgamma(b[i]) - lgamma(a[i] + b[i])
+}
+```
+
+There are two key components to parallelising this loop via ```parallel_map```: an indexing function and a computation/application function.
+
+## Index Function
+The indexing function is the workhorse of the framework, as it allows ```parallel_map``` to manipulate only the variables that are needed by a given thread. This is in contrast to ```reduce_sum```, which will only 'slice' one argument and copy the rest in full to all threads. For the above loop we want to loop over both arguments (```a``` and ```b```), this means that the indexing function will be:
```

Comment:
Can you give more examples of the indexing function? Is this something we should adopt for `reduce_sum` as well?

Perhaps write some Stan psuedo-code? Like parallelize this code:

```
vector[N] mu = ...;
vector[G] effects = ...;
int[N] idxs = ...;
vector[N] y = ...;
vector[N] lp;
for(i in 1:N) {
  lp[i] = normal_lpdf(y[i] | mu[i] + effects[idxs[i]]);
}
```

The problem with the basic `reduce_sum` version of this is that every chunk of work ends up copying all the partials in `effects`. If we could get around that with fancy indexing that would be nice.

Anyway I like the indexing. Indexing is the biggest limiter to `reduce_sum` I think, so if there's away around this that's great.

Reply:
Oh good point on the pseuco-code. I'll update the doc proper later this week, but in the interim, given the Stan model:
```
data {
  int[N] idxs;
  vector[N] y;
}

parameters {
  vector[N] mu;
  vector[G] effects;
  vector<lower=0>[N] sigma;
}

model {
  vector[N] lp;

  for(i in 1:N) {
    lp[i] = normal_lpdf(y[i] | mu[i] + effects[idxs[i]], sigma[i]);
  }
}
```

This would be expressed with ```parallel_map``` via:
```
functions {

  // Return is TBD as need to co-ordinate with Stanc3 to have this
  // transpile to an auto return
  TBD index_fun(int i, F fun, vector y, vector mu, vector effects,
                vector idxs, vector sigma) {
    return fun(y[i], mu[i], effects[idxs[i]], sigma[i]);
  }

  // effects will already have been indexed, so don't
  // need idxs array here
  real apply_fun(vector y, vector mu, vector effects) {
    return normal_lpdf(y | mu + effects, sigma);
  }
}

data {
  int[N] idxs;
  vector[N] y;
  int grainsize;
}

parameters {
  vector[N] mu;
  vector[G] effects;
}

model {
  vector[N] lp;

  parallel_map(apply_fun, index_fun, lp, grainsize,
               y, mu, effects, idxs, sigma);
}
```

Reply:
> Can you give more examples of the indexing function? Is this something we should adopt for reduce_sum as well?

I'm a bit biased because I think the approach is neat, but I think it would expand the use-cases for ```reduce_sum``` by quite a bit. I've had a few situations where I've wanted to use it, but the parameters were too large for the copy-cost to be worth the parallelisation.

Reply:
> but I think it would expand the use-cases for `reduce_sum`

Well I was thinking deprecating reduce sum if this is just better.

The thing `reduce_sum` does that this doesn't is work with ranges so that the parallel work can be larger. Maybe this can work with ranges too?

You'll need to make copies of the shared arguments or do atomic operations on the adjoint accumulates.

I think because the indexing is custom, you'd need compiler help to figure out what to copy. For instance:
```
return fun(y[i], mu[i], effects[idxs[i]], sigma[i]);
```

1. `effects[idxs[i]]` could access the same element of effects at two different `i` s, so there's a race condition on the reverse pass
2. Before evaluating this statement, you won't know what to copy for each chunk of work
3. So somehow I think you'd need to parse this and put compiler stuff in there to know what to copy. Like:

`fun(deep_copy(y[i]), deep_copy(mu[i]), deep_copy(effects[idxs[i]]), deep_copy(sigma[i]))`

Reply:
That's actually the main attraction of the indexing function, that it's used for deep copying the vars and varis.

Going back to the c++ side of things, the indexing function is defined as:
```
auto [&](int i, const auto& fun, const auto& y, const auto& mu,
         const auto& effects, const auto& sigma, const auto& idxs) {
    return fun(y[i], mu[i], effects[idxs[i]], sigma[i]);
  }
```

This is then used to deep copy just the vars at that iteration, using:
```
    auto var_copier = [&](const auto&... xargs) {
      return std::tuple<decltype(deep_copy_vars(xargs))...>(
        deep_copy_vars(xargs)...);
    };

...

    tbb::parallel_for(
      tbb::blocked_range<size_t>(0, S, grainsize),
      [&](const tbb::blocked_range<size_t>& r) {
        // Run nested autodiff in this scope
        const nested_rev_autodiff nested;

        for (size_t i = r.begin(); i < r.end(); ++i) {

        ...
        auto args_tuple_local_copy = index_fun(i, var_copier, x...);
        ...
        }
      });
```

> effects[idxs[i]] could access the same element of effects at two different i s, so there's a race condition on the reverse pass

If the vars are deep copied, is there a race condition?

[Full implementation here](https://github.com/andrjohns/math/blob/b57ac53f1dcff54ee6c7acd3b36b1dda0ff877d7/stan/math/rev/functor/parallel_map.hpp)

Reply:
> If the vars are deep copied, is there a race condition?

Nope! If they're getting deep copied, that's that, you have your own vari and the vari s can be propagated back from there as necessary.

Reply:
Ooo, I see, the `fun` in `index_fun` is deep copying. I thought the fun was the user's function. Yes that makes sense and also the auto return type makes sense now.

Reply:
Yeah it's used in a few contexts:

Counting vars:
```
int nvars = index_fun(0, var_counter, x...);
```

Copying varis:
```
index_fun(i, vari_saver(&vari_map[i]), x...);
```

Deep-copying vars:
```
auto args_tuple_local_copy = index_fun(i, var_copier, x...);
```

Applying function to an iteration (```prim``` only, ```rev``` applies function to copied vars):
```
result[i] = index_fun(i, app_fun, x...);
```